tsdb/importer/import.go

```diff
@@ -0,0 +1,305 @@
+		maxValidTimestamp = maxInt(maxValidTimestamp, currentTimestampMs)
+
+		var lset prom_labels.Labels
+		_ = parser.Metric(&lset)
+
+		currentTime = time.Unix(currentTimestampMs/1000, 0)
+		currentSample := &metricSample{TimestampMs: currentTimestampMs, Value: val, Labels: tsdb_labels.FromMap(lset.Map())}
+
+		if startTime.IsZero() {
+			startTime = currentTime
+			currentBucket = append(currentBucket, currentSample)
+			continue
+		}
+
+		timeDelta := currentTime.Sub(startTime)
+		if timeDelta.Seconds()*1000 >= BlockDuration {
+			startTime = currentTime
+
+			start := int64(startTime.Second() * 1000)
+			end := int64(currentTime.Second() * 1000)
+			blockPath, err := pushToDisk(currentBucket, dbPath, dbMint, dbMaxt, start, end, skipTimestampCheck, logger)
+			if err != nil {
+				if err == OverlappingBlocksError {
+					level.Warn(logger).Log("msg", fmt.Sprintf("could not merge with range %d to %d as it overlaps with target DB", start, end))
+				} else {
+					return nil, err
+				}
+			}
+			blockPaths = append(blockPaths, blockPath)
+			currentBucket = []*metricSample{currentSample}
+		} else {
+			currentBucket = append(currentBucket, currentSample)
+		}
+	}
+	// Last bucket to be added
+	blockPath, err := pushToDisk(currentBucket, dbPath, dbMint, dbMaxt, int64(startTime.Second()*1000), int64(currentTime.Second()*1000), skipTimestampCheck, logger)
+	if err != nil {
+		if err == OverlappingBlocksError {
+			level.Warn(logger).Log("msg", fmt.Sprintf("could not merge with range %d to %d as it overlaps with target DB", int64(startTime.Second()*1000), int64(currentTime.Second()*1000)))
+		} else {
+			return nil, err
+		}
+	}
+	blockPaths = append(blockPaths, blockPath)
+	return blockPaths, nil
+}
+
+// pushToDisk verifies the sample is compatible with the target TSDB instance, and then creates a new block
+// from the sample data.
+func pushToDisk(samples []*metricSample, dbDir string, dbMint, dbMaxt, startTime, endTime timestamp, skipTimestampCheck bool, logger log.Logger) (string, error) {
+	if !skipTimestampCheck {
+		err := verifyIntegration(dbMint, dbMaxt, startTime, endTime)
+		if err != nil {
+			return "", OverlappingBlocksError
+		}
+	}
+	return createBlock(samples, dbDir, logger)
+}
+
+// createHead creates a TSDB writer head to write the sample data to.
+func createHead(samples []*metricSample, chunkRange int64, logger log.Logger) (*tsdb.Head, error) {
+	head, err := tsdb.NewHead(nil, logger, nil, chunkRange)
+	if err != nil {
+		return nil, err
+	}
+	app := head.Appender()
+	for _, sample := range samples {
+		_, err = app.Add(sample.Labels, sample.TimestampMs, sample.Value)
+		if err != nil {
+			return nil, err
+		}
+	}
+	err = app.Commit()
+	if err != nil {
+		return nil, err
+	}
+	return head, nil
+}
+
+// createBlock creates a 2h block from the samples passed to it, and writes it to disk.
+func createBlock(samples []*metricSample, dir string, logger log.Logger) (string, error) {
```

Comment:
I think at this point lets try to expose the existing `createBlock` and `createHead` these are also used in Thanos so they can reuse them when exposed.
cc @bwplotka

Reply:
Makes sense, do you have a suggestion on where to place the exported functions?

Reply:
maybe `tsdb/tsdbutil/block.go`

Reply:
ping @dipack95

Reply:
When I try to move the `createBlock`, `createHead` functions into a new file called `tsdb/tsdbutil/block.go`, during the `make` process, it errors out with 

`ERRO Running error: context loading failed: failed to load program with go/packages: internal error: go list gives conflicting information for package github.com/prometheus/prometheus/tsdb [github.com/prometheus/prometheus/tsdb.test]`

This happens cause I need to import the `tsdb` package to open a new `Head` writer. Any idea on how to solve this?

Reply:
maybe look at how it is done in Thanos

pkg/testutil/prometheus.go